<a href="https://colab.research.google.com/github/Raihan-Islam-Sonet/weather_prediction_project/blob/main/weather_prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Mount the drive and read the dataset**

In [ ]:
from google.colab import drive
drive.mount("/content/gdrive")

Mounted at /content/gdrive


#**Import library for dataset read**

In [ ]:
import numpy as np
import pandas as pd

import sklearn
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

from sklearn.linear_model import LinearRegression

from matplotlib import pyplot as plt
%matplotlib inline

In [ ]:
weather = pd.read_csv('gdrive/My Drive/Colab Notebooks/dataset/weatherHistory.csv')

#**Show the dataset**

In [ ]:
weather.head()

Formatted Date  ...                      Daily Summary
0  2006-04-01 00:00:00.000 +0200  ...  Partly cloudy throughout the day.
1  2006-04-01 01:00:00.000 +0200  ...  Partly cloudy throughout the day.
2  2006-04-01 02:00:00.000 +0200  ...  Partly cloudy throughout the day.
3  2006-04-01 03:00:00.000 +0200  ...  Partly cloudy throughout the day.
4  2006-04-01 04:00:00.000 +0200  ...  Partly cloudy throughout the day.

[5 rows x 12 columns]

#**Show all the columns**

In [ ]:
weather.columns

Index(['Formatted Date', 'Summary', 'Precip Type', 'Temperature (C)',
       'Apparent Temperature (C)', 'Humidity', 'Wind Speed (km/h)',
       'Wind Bearing (degrees)', 'Visibility (km)', 'Loud Cover',
       'Pressure (millibars)', 'Daily Summary'],
      dtype='object')

In [ ]:
weather.shape

(96453, 12)

#**Describe the dataset**

In [ ]:
weather.describe()

Temperature (C)  ...  Pressure (millibars)
count     96453.000000  ...          96453.000000
mean         11.932678  ...           1003.235956
std           9.551546  ...            116.969906
min         -21.822222  ...              0.000000
25%           4.688889  ...           1011.900000
50%          12.000000  ...           1016.450000
75%          18.838889  ...           1021.090000
max          39.905556  ...           1046.380000

[8 rows x 8 columns]

#**Show the null value**

In [ ]:
weather.isnull().any()

Formatted Date              False
Summary                     False
Precip Type                  True
Temperature (C)             False
Apparent Temperature (C)    False
Humidity                    False
Wind Speed (km/h)           False
Wind Bearing (degrees)      False
Visibility (km)             False
Loud Cover                  False
Pressure (millibars)        False
Daily Summary               False
dtype: bool

#**Value counts**

In [ ]:
weather['Precip Type'].value_counts()

rain    85224
snow    10712
Name: Precip Type, dtype: int64

#**Convert string into number**

In [ ]:
weather.loc[weather['Precip Type'].isnull(),'Precip Type'] = 'rain'

In [ ]:
weather.loc[weather['Precip Type'] == 'rain', 'Precip Type'] = 1
weather.loc[weather['Precip Type'] == 'snow', 'Precip Type'] = 0

In [ ]:
weather_num = weather[list(weather.dtypes[weather.dtypes!='object'].index)]

In [ ]:
weather_y = weather_num.pop('Temperature (C)')
weather_X = weather_num

In [ ]:
weather.head()

Formatted Date  ...                      Daily Summary
0  2006-04-01 00:00:00.000 +0200  ...  Partly cloudy throughout the day.
1  2006-04-01 01:00:00.000 +0200  ...  Partly cloudy throughout the day.
2  2006-04-01 02:00:00.000 +0200  ...  Partly cloudy throughout the day.
3  2006-04-01 03:00:00.000 +0200  ...  Partly cloudy throughout the day.
4  2006-04-01 04:00:00.000 +0200  ...  Partly cloudy throughout the day.

[5 rows x 12 columns]

#**Split dataset into train and test**

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(weather_X, weather_y, test_size = 0.2, random_state = 4)

In [ ]:
X_train.head()

Apparent Temperature (C)  Humidity  ...  Loud Cover  Pressure (millibars)
70626                 21.061111      0.31  ...         0.0               1005.87
52457                 25.016667      0.36  ...         0.0               1025.36
90690                  0.738889      0.89  ...         0.0               1014.75
69528                 13.772222      0.78  ...         0.0               1014.56
92419                 23.288889      0.82  ...         0.0               1022.05

[5 rows x 7 columns]

#**Linear Regression**

In [ ]:
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [ ]:
prediction = model.predict(X_test)

In [ ]:
np.mean((prediction-y_test)**2)

0.902274371188337

In [ ]:
pd.DataFrame({'actual': y_test,
              'prediction': prediction,
              'diff': (y_test-prediction)})

actual  prediction      diff
37443  -2.288889   -3.355714  1.066825
86534   8.861111    9.418530 -0.557419
2082    9.805556    9.701321  0.104235
53130  27.222222   27.096837  0.125385
45196  17.705556   17.302053  0.403503
...          ...         ...       ...
55387 -10.066667  -11.960987  1.894320
25976   9.972222    9.831699  0.140523
32423  12.777778   15.436904 -2.659127
26570  15.277778   16.307788 -1.030010
45834   2.200000    3.666659 -1.466659

[19291 rows x 3 columns]

**The error for this model is --- 0.90227**

#**Plynomial Regression**

In [ ]:
from sklearn.preprocessing import PolynomialFeatures

poly = PolynomialFeatures(degree = 4)
X_poly = poly.fit_transform(X_train)

poly.fit(X_poly, y_train)
lin2 = LinearRegression()
lin2.fit(X_poly, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [ ]:
prediction2 = lin2.predict(poly.fit_transform(X_test))
np.mean((prediction2-y_test)**2)

0.14602317508410093

In [ ]:
pd.DataFrame({'actual': y_test,
              'prediction': prediction2,
              'diff': (y_test-prediction2)})

actual  prediction      diff
37443  -2.288889   -2.188155 -0.100734
86534   8.861111    9.113336 -0.252225
2082    9.805556    9.602271  0.203285
53130  27.222222   27.130414  0.091808
45196  17.705556   17.770751 -0.065196
...          ...         ...       ...
55387 -10.066667  -10.172420  0.105753
25976   9.972222    9.773758  0.198464
32423  12.777778   13.294539 -0.516761
26570  15.277778   15.414607 -0.136829
45834   2.200000    2.957100 -0.757100

[19291 rows x 3 columns]

**Error -- 0.14602**

#**Decission Tree Regression**

In [ ]:
from sklearn.tree import DecisionTreeRegressor
regressor = DecisionTreeRegressor(random_state = 0)
regressor.fit(X_train, y_train)

DecisionTreeRegressor(ccp_alpha=0.0, criterion='mse', max_depth=None,
                      max_features=None, max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, presort='deprecated',
                      random_state=0, splitter='best')

In [ ]:
prediction3 = regressor.predict(X_test)
np.mean((prediction3-y_test)**2)

0.006723574480775507

In [ ]:
pd.DataFrame({'actual': y_test,
              'prediction': prediction3,
              'diff': (y_test-prediction3)})

actual  prediction          diff
37443  -2.288889   -2.272222 -1.666667e-02
86534   8.861111    8.855556  5.555556e-03
2082    9.805556    9.788889  1.666667e-02
53130  27.222222   27.222222 -3.552714e-15
45196  17.705556   17.705556 -7.105427e-15
...          ...         ...           ...
55387 -10.066667  -10.000000 -6.666667e-02
25976   9.972222    9.972222  0.000000e+00
32423  12.777778   12.777778 -4.263256e-14
26570  15.277778   15.277778  3.552714e-15
45834   2.200000    2.200000  4.440892e-16

[19291 rows x 3 columns]

Error -- 0.0067

#**Random Forest with maximum Dept - 10**

In [ ]:
from sklearn.ensemble import RandomForestRegressor
regr = RandomForestRegressor(max_depth = 10, random_state = 0, n_estimators = 100)
regr.fit(X_train, y_train)

RandomForestRegressor(bootstrap=True, ccp_alpha=0.0, criterion='mse',
                      max_depth=10, max_features='auto', max_leaf_nodes=None,
                      max_samples=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      n_estimators=100, n_jobs=None, oob_score=False,
                      random_state=0, verbose=0, warm_start=False)

In [ ]:
prediction4 = regr.predict(X_test)
np.mean((prediction4-y_test)**2)

0.0071767589236793015

In [ ]:
pd.DataFrame({'actual': y_test,
              'prediction': prediction4,
              'diff': (y_test-prediction4)})

actual  prediction      diff
37443  -2.288889   -2.366102  0.077213
86534   8.861111    8.892260 -0.031149
2082    9.805556    9.847246 -0.041690
53130  27.222222   27.202118  0.020104
45196  17.705556   17.713716 -0.008160
...          ...         ...       ...
55387 -10.066667  -10.100922  0.034255
25976   9.972222    9.913953  0.058269
32423  12.777778   12.781718 -0.003940
26570  15.277778   15.275648  0.002130
45834   2.200000    2.187450  0.012550

[19291 rows x 3 columns]

**Error -- 0.0072**

#**Random Forest with maximum Dept - 50**

In [ ]:
regr2 = RandomForestRegressor(max_depth = 50, random_state = 0, n_estimators = 100)
regr2.fit(X_train, y_train)

RandomForestRegressor(bootstrap=True, ccp_alpha=0.0, criterion='mse',
                      max_depth=50, max_features='auto', max_leaf_nodes=None,
                      max_samples=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      n_estimators=100, n_jobs=None, oob_score=False,
                      random_state=0, verbose=0, warm_start=False)

In [ ]:
prediction5 = regr.predict(X_test)
np.mean((prediction5-y_test)**2)

0.0071767589236793015

In [ ]:
pd.DataFrame({'actual': y_test,
              'prediction': prediction5,
              'diff': (y_test-prediction5)})

actual  prediction      diff
37443  -2.288889   -2.366102  0.077213
86534   8.861111    8.892260 -0.031149
2082    9.805556    9.847246 -0.041690
53130  27.222222   27.202118  0.020104
45196  17.705556   17.713716 -0.008160
...          ...         ...       ...
55387 -10.066667  -10.100922  0.034255
25976   9.972222    9.913953  0.058269
32423  12.777778   12.781718 -0.003940
26570  15.277778   15.275648  0.002130
45834   2.200000    2.187450  0.012550

[19291 rows x 3 columns]